examples/multiturn_demo/agent_loop/mcp_agent_loop.py

```diff
@@ -0,0 +1,201 @@
+                session_id, async_op=True
+            ).async_wait()
+            assert response.success
+        return session_id
+
+    async def tool_session_release(self, tool_worker_name, session_id) -> str | dict:
+        tool_channel_info = self.tool_channel_info[tool_worker_name]
+        await (
+            tool_channel_info["input_channel"]
+            .put(
+                ToolChannelRequest(session_id=session_id, request_type="session_end"),
+                async_op=True,
+            )
+            .async_wait()
+        )
+        response: ToolChannelResponse = await self.tool_worker_output_channel.get(
+            session_id, async_op=True
+        ).async_wait()
+        assert response.success
+
+    async def atool_call(
+        self, generate_context, tool_request: ToolRequest
+    ) -> ToolResponse:
+        tool_name, tool_args = tool_request.name, tool_request.arguments
+        tool_channel_info = self.tool_channel_info[self.tool_name_map[tool_name]]
+        tool_input_channel = tool_channel_info["input_channel"]
+        session_id = await self.tool_session_get(generate_context, tool_name)
+        await tool_input_channel.put(
+            ToolChannelRequest(
+                session_id=session_id,
+                request_type="execute",
+                tool_name=tool_name,
+                tool_args=tool_args,
+            ),
+            async_op=True,
+        ).async_wait()
+        response: ToolChannelResponse = await self.tool_worker_output_channel.get(
+            session_id, async_op=True
+        ).async_wait()
+        assert response.success
+        if isinstance(response.result, (list, dict)):
+            result_text = json.dumps(response.result)
+        else:
+            result_text = str(response.result)
+        return ToolResponse(
+            text=result_text,
+        )
+
+    async def extract_tool_calls(self, response_text) -> tuple[str, list[ToolRequest]]:
+        # random tool call
+        return_function_calls = random.choice(
+            [
+                [
+                    ToolRequest(
+                        name="write_file",
+                        arguments={
+                            "path": "/projects/test/mcp_written.txt",
+                            "content": f"Written by mcp at {time.strftime('%Y-%m-%d %H:%M:%S')}",
+                        },
+                    )
+                ],
+                [
+                    ToolRequest(
+                        name="list_directory", arguments={"path": "/projects/test"}
+                    )
+                ],
+            ]
+        )
+
+        return response_text, return_function_calls
+
+    async def run_one_query(self, prompt_ids: list[int]) -> AgentLoopOutput:
+        orig_prompt_ids = copy.deepcopy(prompt_ids)
+        generate_context: dict[str, Any] = self.generate_context_create()
+        try:
+            for _ in range(5):
```

Comment:
please add comments to explain that 5 is for demo purpose, users can write their own multi-turn control loop logic.